lib/AST/ASTWalker.cpp

```diff
@@ -398,11 +398,12 @@ class Traversal : public ASTVisitor<Traversal, Expr*, Stmt*,
     }
 
     if (auto *rawLiteralExpr = ED->getRawValueUnchecked()) {
-      Expr *newRawExpr = doIt(rawLiteralExpr);
-      if (auto newRawLiteralExpr = dyn_cast<LiteralExpr>(newRawExpr))
-        ED->setRawValueExpr(newRawLiteralExpr);
-      else
-        return true;
+      if (Expr *newRawExpr = doIt(rawLiteralExpr)) {
```

Comment:
Just going by the surrounding code, are we meant to return true to stop reversal when doIt returns nullptr? If we are I guess the outer if would `else return true` and the inner if would ... else unreachable or assert because that means the subclass is badly behaved?

Reply:
Most code appears to be returning `false` (or otherwise falling through to other code) if `doIt` returns `nullptr`.  For example, 

```c++
 bool visitAbstractTypeParamDecl(AbstractTypeParamDecl *TPD) {
    for (auto Inherit: TPD->getInherited()) {
      if (doIt(Inherit))
        return true;
    }
    
    if (auto *ATD = dyn_cast<AssociatedTypeDecl>(TPD)) {
      if (auto *WhereClause = ATD->getTrailingWhereClause()) {
        for (auto &Req: WhereClause->getRequirements()) {
          if (doIt(Req))
            return true;
        }
      }
    }
    return false;
  }
```

That's what I was following.  For the inner if, are you suggesting there is an invariant being broken?

Reply:
I think doIt() given a TypeLoc or Requirement like above returns a boolean where true means bail. doIt() given an Expression returns a replacement expression or nullptr to bail.

Reply:
And yes about the invariant. doIt calls walkToExprPre() which may return a different expression to the one it was given. If it's visiting the raw value of an enum element though it's probably a bug in the client walker if it tells us to replace it with something other than a literal expression.

Reply:
> I think doIt() given a TypeLoc or Requirement like above returns a boolean where true means bail. doIt() given an Expression returns a replacement expression or nullptr to bail.

You're right,  thanks!  I'll fix.

> it's probably a bug in the client walker if it tells us to replace it with something other than a literal expression.

Okay, I'll try that.